File: threads/src/main/java/im/threads/business/database/DatabaseHolder.java
package im.threads.business.database;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.List;

import im.threads.business.models.ChatItem;
import im.threads.business.models.ConsultInfo;
import im.threads.business.models.ConsultPhrase;
import im.threads.business.models.FileDescription;
import im.threads.business.models.MessageState;
import im.threads.business.models.SpeechMessageUpdate;
import im.threads.business.models.Survey;
import im.threads.business.models.UserPhrase;
import im.threads.business.config.BaseConfig;
import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

public final class DatabaseHolder {

    private static DatabaseHolder instance;
    private final ThreadsDbHelper mMyOpenHelper;

    private DatabaseHolder() {
        mMyOpenHelper = new ThreadsDbHelper(BaseConfig.instance.context);
    }

    @NonNull
    public static DatabaseHolder getInstance() {
        if (instance == null) {
            instance = new DatabaseHolder();
        }
        return instance;
    }

    // ChatItems

    /**
     * Nullify instance. For Autotests purposes
     */
    static void eraseInstance() {
        instance = null;
    }

    public void cleanDatabase() {
        mMyOpenHelper.cleanDatabase();
    }

    @NonNull
    public List<ChatItem> getChatItems(int offset, int limit) {
        return mMyOpenHelper.getChatItems(offset, limit);
    }

    @Nullable
    public ChatItem getChatItem(String messageUuid) {
        return mMyOpenHelper.getChatItem(messageUuid);
    }

    public void putChatItems(final List<ChatItem> items) {
        mMyOpenHelper.putChatItems(items);
    }

    public boolean putChatItem(ChatItem chatItem) {
        return mMyOpenHelper.putChatItem(chatItem);
    }

    // FileDescriptions

    public Single<List<FileDescription>> getAllFileDescriptions() {
        return Single.fromCallable(mMyOpenHelper::getAllFileDescriptions)
                .subscribeOn(Schedulers.io());
    }

    // UserPhrase

    public void updateFileDescription(@NonNull FileDescription fileDescription) {
        mMyOpenHelper.updateFileDescription(fileDescription);
    }

    @Nullable
    public ConsultInfo getConsultInfo(@NonNull String id) {
        return mMyOpenHelper.getLastConsultInfo(id);
    }

    public List<UserPhrase> getUnsendUserPhrase(int count) {
        return mMyOpenHelper.getUnsendUserPhrase(count);
    }

    // ConsultPhrase

    public void setStateOfUserPhraseByProviderId(String providerId, MessageState messageState) {
        mMyOpenHelper.setUserPhraseStateByProviderId(providerId, messageState);
    }

    public Single<ConsultPhrase> getLastConsultPhrase() {
        return Single.fromCallable(mMyOpenHelper::getLastConsultPhrase)
                .subscribeOn(Schedulers.io());
    }

    public Completable setAllConsultMessagesWereRead() {
        return Completable.fromCallable(mMyOpenHelper::setAllConsultMessagesWereRead)
                .subscribeOn(Schedulers.io());
    }

    public void setMessageWasRead(String uuid) {
        mMyOpenHelper.setMessageWasRead(uuid);
    }

    public void saveSpeechMessageUpdate(SpeechMessageUpdate speechMessageUpdate) {
        mMyOpenHelper.speechMessageUpdated(speechMessageUpdate);
    }

    // Survey
    @Nullable
    public Survey getSurvey(long sendingId) {
        return mMyOpenHelper.getSurvey(sendingId);
    }

    public Completable setNotSentSurveyDisplayMessageToFalse() {
        return Completable.fromCallable(mMyOpenHelper::setNotSentSurveyDisplayMessageToFalse)
                .subscribeOn(Schedulers.io());
    }

    public Completable setOldRequestResolveThreadDisplayMessageToFalse() {
        return Completable.fromCallable(mMyOpenHelper::setOldRequestResolveThreadDisplayMessageToFalse)
                .subscribeOn(Schedulers.io());
    }

    // Messages
    public int getMessagesCount() {
        return mMyOpenHelper.getMessagesCount();
    }

    // let the DB time to write the incoming message
    public int getUnreadMessagesCount() {
        return mMyOpenHelper.getUnreadMessagesCount();
    }

    @NonNull
    public List<String> getUnreadMessagesUuid() {
        return mMyOpenHelper.getUnreadMessagesUuid();
    }

    /**
     * For Autotests purposes
     *
     * @return MyOpenHelper instance
     */
    ThreadsDbHelper getMyOpenHelper() {
        return mMyOpenHelper;
    }

}

File: threads/src/main/java/im/threads/internal/secureDatabase/DatabaseHolder.kt
package im.threads.internal.secureDatabase

import im.threads.internal.Config
import im.threads.internal.model.ChatItem
import im.threads.internal.model.ConsultInfo
import im.threads.internal.model.ConsultPhrase
import im.threads.internal.model.FileDescription
import im.threads.internal.model.MessageState
import im.threads.internal.model.SpeechMessageUpdate
import im.threads.internal.model.Survey
import im.threads.internal.model.UserPhrase
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

class DatabaseHolder private constructor() {
    /**
     * For Autotests purposes
     *
     * @return MyOpenHelper instance
     */
    val myOpenHelper = ThreadsDbHelper(Config.instance.context)

    fun cleanDatabase() {
        myOpenHelper.cleanDatabase()
    }

    fun getChatItems(offset: Int, limit: Int): List<ChatItem> =
        myOpenHelper.getChatItems(offset, limit)

    fun getChatItem(messageUuid: String?): ChatItem? = myOpenHelper.getChatItem(messageUuid)

    fun putChatItems(items: List<ChatItem?>?) {
        myOpenHelper.putChatItems(items)
    }

    fun putChatItem(chatItem: ChatItem?): Boolean = myOpenHelper.putChatItem(chatItem)

    // FileDescriptions
    val allFileDescriptions: Single<List<FileDescription?>?>
        get() = Single.fromCallable { myOpenHelper.getAllFileDescriptions() }
            .subscribeOn(Schedulers.io())

    // UserPhrase
    fun updateFileDescription(fileDescription: FileDescription) {
        myOpenHelper.updateFileDescription(fileDescription)
    }

    fun getConsultInfo(id: String): ConsultInfo? = myOpenHelper.getLastConsultInfo(id)

    fun getUnsendUserPhrase(count: Int): List<UserPhrase> = myOpenHelper.getUnsendUserPhrase(count)

    // ConsultPhrase
    fun setStateOfUserPhraseByProviderId(providerId: String?, messageState: MessageState?) {
        myOpenHelper.setUserPhraseStateByProviderId(providerId, messageState)
    }

    val lastConsultPhrase: Single<ConsultPhrase?> =
        Single.fromCallable { myOpenHelper.getLastConsultPhrase() }
            .subscribeOn(Schedulers.io())

    fun setAllConsultMessagesWereRead(): Completable {
        return Completable.fromCallable { myOpenHelper.setAllConsultMessagesWereRead() }
            .subscribeOn(Schedulers.io())
    }

    fun setMessageWasRead(uuid: String?) {
        uuid?.let { myOpenHelper.setMessageWasRead(it) }
    }

    fun saveSpeechMessageUpdate(speechMessageUpdate: SpeechMessageUpdate?) {
        myOpenHelper.speechMessageUpdated(speechMessageUpdate)
    }

    fun getSurvey(sendingId: Long): Survey? {
        return myOpenHelper.getSurvey(sendingId)
    }

    fun setNotSentSurveyDisplayMessageToFalse(): Completable {
        return Completable.fromCallable { myOpenHelper.setNotSentSurveyDisplayMessageToFalse() }
            .subscribeOn(Schedulers.io())
    }

    fun setOldRequestResolveThreadDisplayMessageToFalse(): Completable {
        return Completable.fromCallable { myOpenHelper.setOldRequestResolveThreadDisplayMessageToFalse() }
            .subscribeOn(Schedulers.io())
    }

    val messagesCount: Int = myOpenHelper.getMessagesCount()

    val unreadMessagesCount: Int = myOpenHelper.getUnreadMessagesCount()

    val unreadMessagesUuid: List<String?> = myOpenHelper.getUnreadMessagesUuid()

    fun checkAndUpdate() {
        val oldHelper = im.threads.internal.database.ThreadsDbHelper(Config.instance.context)
        if (needMigrateToNewDB(oldHelper)) {
            putChatItems(oldHelper.getChatItems(0, -1))
            myOpenHelper.putFileDescriptions(oldHelper.getAllFileDescriptions())
            oldHelper.cleanDatabase()
        }
    }

    fun needMigrateToNewDB(helper: im.threads.internal.database.ThreadsDbHelper): Boolean {
        return helper.getChatItems(0, -1).size > 0 || helper.getAllFileDescriptions().size > 0
    }

    companion object {

        private var instance: DatabaseHolder? = null

        @JvmStatic
        fun getInstance(): DatabaseHolder {
            if (instance == null) {
                instance = DatabaseHolder()
                instance?.checkAndUpdate()
            }
            return instance!!
        }
        // ChatItems
        /** Nullify instance. For Autotests purposes */
        fun eraseInstance() {
            instance = null
        }
    }
}
